Support for i18n

onli opened the pull request:
See the discussion in #116.

This implements i18n via gettext, using the rust gettext module from https://github.com/justinas/gettext. A simple makefile is provided to help create the .mo files which the program reads. Translations are given as variables into the templates, but are strings in the sourcecode – that is not perfect, but https://github.com/nickel-org/rust-mustache does not support the better approach of calling a function from within the template.

I tested this till the point of sending the email out.

callahad commented on src/handlers/oidc.rs:
```diff
@@ -115,21 +117,33 @@ broker_handler!(Auth, |app, req| {
+            "lang/german.mo"
+        } else {
+            "lang/english.mo" // fallback, for easier catalog creation below
+        };
```
Is there a way to avoid hard-coding each language?

onli replied:
Hm. We could probably try to select a translation file based on the langtag-language?

callahad commented on src/handlers/oidc.rs:
```diff
@@ -115,21 +117,33 @@ broker_handler!(Auth, |app, req| {
 
     } else {
+        let langtag_de: LanguageTag = "de".parse().unwrap();
+        let accept_language = req.headers.get::<AcceptLanguage>().expect("No accepted language set");
```
Fall back to English when no AcceptLanguage header is passed

onli replied:
https://github.com/portier/portier-broker/pull/124/commits/52f83cd2b4dc5116951d1f9793e4c422e7602848 does that.

callahad commented on src/handlers/oidc.rs:
```diff
@@ -115,21 +117,33 @@ broker_handler!(Auth, |app, req| {
+                               ("title", &catalog.gettext("Confirm your address")),
+                               ("explanation", catalog.gettext("We've sent you an email to confirm your address.")),
+                               ("use", catalog.gettext("Use the link in that email to login to")),
+                               ("alternate", catalog.gettext("Alternatively, enter the code from the email to continue in this browser tab:")),
```
I wish there was a way to distinguish between variables (`client_id` and `session_id`) and translated text in the templates. Is there a gettext helper for our templating system? Can we pass in the whole catalog and thus have something that looks like namespaces?

onli replied:
If we want to do things like that, I think we need to switch from mustache to handlebar or another template system. Is that something we want to tackle now?

callahad commented on tmpl/email_html.mustache:
```diff
@@ -3,21 +3,20 @@
       <p style="margin: 24px; font:normal 1.25em sans-serif">
-        You received this email so that we may confirm your email address
-        and finish your login to: <em>{{ client_id }}</em>
+        {{ explanation }} <em>{{ client_id }}</em>
```
This would break for RTL languages... any idea of how to fix it? Not a blocker until we get our first contributed RTL translation. :)

onli replied:
The normal solution would be to give the variable to the gettext helper, with the translation placing the variable content where it belongs. Would be ugly here because of the html, but that can be ignore if necessary, but… frankly, I'm not sure whether https://docs.rs/gettext/0.2.0/gettext/ supports it, I don't see it clearly in the docs.

onli commented:
Short update: https://github.com/portier/portier-broker/pull/124/commits/52f83cd2b4dc5116951d1f9793e4c422e7602848 addresses the the comments I could address, namely the fallback to English if no language header is sent. I also tested that, it works. That commit also improves the code to actually go through all the language headers, properly in the order it is sorted on the client, and using the first found language it supports. That includes a basis to add more languages without bigger code changes, only the arrays linking language header to catalogue file would need an update if adding a new language (there might be a nicer language there, but this should work okay).

I asked @stephank to go through the code, as I again was fighting Rust trying to get this to work.

onli commented:
@callahad, @stephank ?

stephank commented on src/handlers/oidc.rs:
```diff
@@ -115,21 +118,48 @@ broker_handler!(Auth, |app, req| {
+        }
+        
+        let f = File::open(lang_file).expect("could not open the catalog");
+        let catalog = Catalog::parse(f).expect("could not parse the catalog");
```
I'll repeat the comments from our private discussion on IRC here:

 - For consistency, use `to_string` instead of `to_owned`. (Converting from `&str` to `String`.) Not sure what the Rust preferred option is, but that's what we've used everywhere.

 - LanguageTag implements the `Clone` trait, which'd be a better option than two `langtag_en` variables.

 - I'd opt for keeping the language file names the same as the language tag, which could then simply be mapped to `lang/{}.mo`. We'd still probably want to hardcode the list of files in the code, but it'd be one less 'term' to deal with.

 - We should probably parse catalogs once on init, rather than on every request.

 - Feels like creating a default header value is a roundabout way of just initialising `lang_file` to the default language? When the for-loop falls through, it'd simply use that.

Here's a gist of how I'd write it: https://gist.github.com/stephank/a6c4647d16956335d2f48a62662362f8

onli replied:
Thanks a lot @stephank. I pushed https://github.com/portier/portier-broker/pull/124/commits/7a43105d628d3f11f94b0347d0fb0360e23401a4, which is a mix of our two approaches. I'm very happy about the improvements you could show me.

To your comments:

 * I changed that to to_owned, but later on was able to remove that part completely
 * `Clone` is way better here, I searched that
 * Agreed and changed
 * We probably should, but I don't know how to implement that. Have a static(?) Vec<LanguageTag, Catalog> in config that is filled on init? That's too much rust for me. I instead adapted your code to only create one catalogue, the needed one.
 * Yes, that was pretty unnecessary. I did not understand how to use `if let Some(accept)`.

Could we merge this now? Or does someone want to move the catalogue parsing to init? I think that would be nice, but not something that should block this PR if no one wants to do it immediately. We can open an issue for  that.

stephank commented:
Pushed the following changes:

 - Squashed into one commit
 - Rebased on master
 - Translated email subject
 - Moved catalog parsing to config
 - Added a Dutch translation
 - Moved Makefile logic to `build.rs`
 - Updated `docker-build.sh` to package compiled catalogs

stephank commented:
Alright, clippy is now happy. I'll let you have one more look at the changes I've made, @onli, and if you think it's okay, you can hit that merge button! :)

onli commented:
Thanks, great!
